test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1

```diff
@@ -143,6 +142,143 @@ function GetTestData
     return $body
 }
 
+<#
+    Extract the raw content from a response and
+    convert the embedded json to PSObjects.
+    Assumes the raw content is from the  HTTPListener.
+#>
+function Convert-ResponseRawContent
+{
+    param
+    (
+        [Parameter(Mandatory)]
+        [ValidateNotNull()]
+        [Microsoft.PowerShell.Commands.WebResponseObject] $Response
+    )
+
+    $index = $Response.RawContent.IndexOf("{")
+    [string] $rawContent = $Response.RawContent.SubString($index)
+    return ($rawContent | ConvertFrom-Json)
+}
+
+function ExecuteRedirectRequest
+{
+    param (
+        [Parameter(Mandatory)]
+        [string]
+        $uri,
+
+        [switch] $UsePost,
+
+        [switch] $PreserveAuthorizationOnRedirect
+    )
+    $result = [PSObject]@{Output = $null; Error = $null; Content = $null}
+
+    try
+    {
+        $headers = @{"Authorization" = "test"}
+        if ($UsePost)
+        {
+            $result.Output = Invoke-WebRequest -Uri $uri -TimeoutSec 5 -Headers $headers -PreserveAuthorizationOnRedirect:$PreserveAuthorizationOnRedirect.IsPresent -Method POST
+        }
+        else
+        {
+            $result.Output = Invoke-WebRequest -Uri $uri -TimeoutSec 5 -Headers $headers -PreserveAuthorizationOnRedirect:$PreserveAuthorizationOnRedirect.IsPresent
+        }
+
+        $result.Content = Convert-ResponseRawContent -Response $result.Output
+    }
+    catch
+    {
+        $result.Error = $_
+    }
+
+    return $result
+}
+
+Describe "Invoke-WebRequest redirect tests" -Tags "Feature" {
+
+    BeforeAll {
+        $null = Start-HttpListener -AsJob -Port 8082
+    }
+
+    AfterAll {
+        $null = Stop-HttpListener -Port 8082
+    }
+
+    <#
+        Defines the list of redirect codes to test as well as the
+        expected Method when the redirection is handled.
+        See https://msdn.microsoft.com/en-us/library/windows/apps/system.net.httpstatuscode(v=vs.105).aspx 
+        for additonal details.
+    #>
+    $redirectTests = @(
+      @{redirectType = 'MultipleChoices'; redirectedMethod='POST'}
+      @{redirectType = 'Ambiguous'; redirectedMethod='POST'} # Synonym for MultipleChoices
+
+      @{redirectType = 'Moved'; redirectedMethod='GET'}
+      @{redirectType = 'MovedPermanently'; redirectedMethod='GET'} # Synonym for Moved
+
+      @{redirectType = 'Found'; redirectedMethod='GET'}
+      @{redirectType = 'Redirect'; redirectedMethod='GET'} # Synonym for Found
+
+      @{redirectType = 'redirectMethod'; redirectedMethod='GET'}
+      @{redirectType = 'SeeOther'; redirectedMethod='GET'} # Synonym for RedirectMethod
+
+      @{redirectType = 'TemporaryRedirect'; redirectedMethod='GET'}
+      @{redirectType = 'RedirectKeepVerb'; redirectedMethod='GET'} # Synonym for TemporaryRedirect
+    )
+
+    It "Validates Invoke-WebRequest with -PreserveAuthorizationOnRedirect preserves the authorization header on redirect" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
```

Comment:
$response.Content is a byte array, not string ($result.Output.Content is this byte array).  $response.RawContent has a few lines of text inserted by the underlying API followed by the json form of the original request; inserted by the httplistener.  The purpose of the function is to extract the json from the RawContent and return it explicitly since it is needed to verify the incoming request.